cep47-logic/src/lib.rs

```diff
@@ -178,46 +187,25 @@ pub trait CEP47Contract<Storage: CEP47Storage>: WithStorage<Storage> {
         if self.is_paused() {
             return Err(Error::PermissionDenied);
         }
-        let mut sender_tokens = self.storage().get_tokens(sender);
 
         for token_id in token_ids.iter() {
-            if !sender_tokens.contains(token_id) {
-                return Err(Error::PermissionDenied);
+            let token_owner = self.storage().owner_of(token_id);
+            if let Some(token_owner) = token_owner {
+                if &token_owner != sender {
+                    return Err(Error::NotAnOwner);
+                }
             }
-            sender_tokens.retain(|x| x != token_id);
-        }
-        let mut recipient_tokens = self.storage().get_tokens(recipient);
-        recipient_tokens.append(&mut token_ids.clone());
-        self.storage_mut().set_tokens(sender, sender_tokens);
-        self.storage_mut().set_tokens(recipient, recipient_tokens);
-
-        // Emit transfer event.
-        self.storage_mut().emit(CEP47Event::Transfer {
-            sender: *sender,
-            recipient: *recipient,
-            token_ids: token_ids.clone(),
-        });
-        Ok(())
-    }
-
-    fn transfer_all_tokens(&mut self, sender: &Key, recipient: &Key) -> Result<(), Error> {
-        if self.is_paused() {
-            return Err(Error::PermissionDenied);
         }
-        let mut sender_tokens = self.storage().get_tokens(sender);
-        let mut recipient_tokens = self.storage().get_tokens(recipient);
-        recipient_tokens.append(&mut sender_tokens);
 
-        self.storage_mut().set_tokens(sender, sender_tokens.clone());
-        self.storage_mut().set_tokens(recipient, recipient_tokens);
+        self.storage_mut()
+            .transfer_many(sender, recipient, token_ids);
 
         // Emit transfer event.
         self.storage_mut().emit(CEP47Event::Transfer {
             sender: *sender,
             recipient: *recipient,
-            token_ids: sender_tokens,
+            token_ids: token_ids.clone(),
```

Comment:
We should be able to get rid of `.clone()` here, right? It can be moved into `Transfer` since it's the last usage.

Reply:
Not really as token_ids is reference. That would require changing more code. Can do in subsequent PR.